File: trxy/Models/Document.swift
//
//  Document.swift
//  trxy
//
//  Created by Adam Anderson on 1/17/20.
//  Copyright © 2020 mtusk. All rights reserved.
//

import Cocoa
import SwiftUI

class Document: NSDocument {
    
    // references:
    // - https://www.raywenderlich.com/1809473-uidocument-from-scratch
    // - https://www.techotopia.com/index.php/Managing_Files_using_the_iOS_8_UIDocument_Class
    
    @objc var content = Content(testResults: [TestResult]())
    var contentViewController: ViewController?
    
    override init() {
        super.init()
        // Add your subclass-specific initialization here.
    }

    override class var autosavesInPlace: Bool {
        return false
    }

    override func makeWindowControllers() {
        // Returns the storyboard that contains your document window.
        let storyboard = NSStoryboard(name: NSStoryboard.Name("Main"), bundle: nil)
        if let windowController = storyboard.instantiateController(withIdentifier: NSStoryboard.SceneIdentifier("Document Window Controller")) as? NSWindowController {
            addWindowController(windowController)
            
            // Set the view controller's represented object as your document.
            if let contentVC = windowController.contentViewController as? ViewController {
                contentVC.representedObject = content
                self.contentViewController = contentVC
            }
        }
    }

    override func data(ofType typeName: String) throws -> Data {
        // Insert code here to write your document to data of the specified type, throwing an error in case of failure.
        // Alternatively, you could remove this method and override fileWrapper(ofType:), write(to:ofType:), or write(to:ofType:for:originalContentsURL:) instead.
        throw NSError(domain: NSOSStatusErrorDomain, code: unimpErr, userInfo: nil)
    }

    override func read(from data: Data, ofType typeName: String) throws {
        content.read(from: data)
    }
}

File: trxy.MacOS/Models/Document.cs
using System;
using System.Collections.Generic;
using AppKit;
using Foundation;
using trxy.MacOS.Models;

namespace trxy.MacOS
{
    [Register("Document")]
    public class Document : NSDocument
    {
        private Content content = new Content(new List<TestResult>());
        private ViewController viewController;

        public Document(IntPtr handle) : base(handle)
        {
            // Add your subclass-specific initialization here.
        }

        public override void WindowControllerDidLoadNib(NSWindowController windowController)
        {
            base.WindowControllerDidLoadNib(windowController);
            // Add any code here that needs to be executed once the windowController has loaded the document's window.
        }

        [Export("autosavesInPlace")]
        public static bool AutosaveInPlace()
        {
            return true;
        }

        public override void MakeWindowControllers()
        {
            // Override to return the Storyboard file name of the document.
            var storyboard = NSStoryboard
                .FromName("Main", null)
                .InstantiateControllerWithIdentifier("Document Window Controller");

            var windowController = (NSWindowController)storyboard;

            this.AddWindowController(windowController);

            if (windowController.ContentViewController is ViewController)
            {
                this.viewController = windowController.ContentViewController as ViewController;
                this.viewController.RepresentedObject = this.content;
            }
        }

        public override NSData GetAsData(string typeName, out NSError outError)
        {
            // Insert code here to write your document to data of the specified type. 
            // If outError != NULL, ensure that you create and set an appropriate error when returning nil.
            throw new NotImplementedException();
        }

        public override bool ReadFromData(NSData data, string typeName, out NSError outError)
        {
            outError = null;

            this.content.Read(data);

            return true;
        }
    }
}
